File: src/main/java/seremis/geninfusion/soul/TraitRegistry.java
package seremis.geninfusion.soul;

import seremis.geninfusion.api.soul.ITrait;
import seremis.geninfusion.api.soul.ITraitRegistry;

import java.util.*;

public class TraitRegistry implements ITraitRegistry {

    public LinkedHashMap<String, ITrait> traits = new LinkedHashMap<String, ITrait>();
    public LinkedHashMap<ITrait, String> names = new LinkedHashMap<ITrait, String>();
    public LinkedHashMap<ITrait, Integer> ids = new LinkedHashMap<ITrait, Integer>();

    public LinkedHashMap<ITrait, List<String>> methods = new LinkedHashMap<ITrait, List<String>>();

    @Override
    public void registerTrait(String name, ITrait trait) {
        if(!traits.containsKey(name)) {
            traits.put(name, trait);
            names.put(trait, name);
            ids.put(trait, ids.size());
        }
    }

    @Override
    public void unregisterTrait(ITrait trait) {
        traits.remove(getName(trait));
        names.remove(trait);
        ids.remove(trait);
    }

    @Override
    public void unregisterTrait(String name) {
        unregisterTrait(getTrait(name));
    }

    @Override
    public ITrait getTrait(String name) {
        return traits.containsKey(name) ? traits.get(name) : null;
    }

    @Override
    public String getName(ITrait trait) {
        return names.containsKey(trait) ? names.get(trait) : null;
    }

    @Override
    public int getId(String name) {
        return ids.get(getTrait(name));
    }

    @Override
    public int getId(ITrait trait) {
        return getId(getName(trait));
    }

    @Override
    public LinkedList<ITrait> getTraits() {
        return new LinkedList<ITrait>(traits.values());
    }
}

File: src/main/scala/seremis/geninfusion/soul/TraitRegistry.scala
package seremis.geninfusion.soul

import seremis.geninfusion.api.soul.{ITrait, ITraitRegistry}

class TraitRegistry extends ITraitRegistry {

    var traits: Map[String, ITrait] = Map()

    var names: Map[ITrait, String] = Map()

    var ids: Map[ITrait, Integer] = Map()

    override def registerTrait(name: String, trt: ITrait) {
        if (!traits.contains(name)) {
            traits += (name -> trt)
            names += (trt -> name)
            ids += (trt -> ids.size)
        }
    }

    override def unregisterTrait(trt: ITrait) {
        traits -= getName(trt)
        names -= trt
        ids -= trt
    }

    override def unregisterTrait(name: String) {
        unregisterTrait(getTrait(name))
    }

    override def getTrait(name: String): ITrait = {
        if (traits.contains(name)) traits.get(name).get else null
    }

    override def getName(trt: ITrait): String = {
        if (names.contains(trt)) names.get(trt).get else null
    }

    override def getId(name: String): Int = ids.get(getTrait(name)).get

    override def getId(trt: ITrait): Int = getId(getName(trt))

    override def getTraits: Array[ITrait] = {
        traits.values.toArray
    }
}